clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java

```diff
@@ -207,6 +226,232 @@ public NetworkClientDelegate.UnsentRequest toUnsentRequest() {
+     *
+     * <p>If the request is new. It will be enqueued and will be sent upon the next
+     * poll.
+     *
+     * <p>If the same request has been sent, the request's {@code CompletableFuture} will be completed upon the
+     * completion of the existing one.
+     *
+     * <p>There is a special handling for the {@link UnsentOffsetFetchRequest}. If a duplicated request was sent
+     * previously, we will chain the future to the current future.
+     */
+
+    class PendingRequests {
+        // Queue is used to ensure the sequence of commit
+        Queue<UnsentOffsetCommitRequest> unsentOffsetCommits = new LinkedList<>();
+        List<UnsentOffsetFetchRequest> unsentOffsetFetches = new ArrayList<>();
+        List<UnsentOffsetFetchRequest> inflightOffsetFetches = new ArrayList<>();
+
+        public boolean hasUnsentRequests() {
+            return !unsentOffsetCommits.isEmpty() || !unsentOffsetFetches.isEmpty();
+        }
+
+        public CompletableFuture<ClientResponse> addOffsetCommitRequest(final Map<TopicPartition, OffsetAndMetadata> offsets) {
+            UnsentOffsetCommitRequest request = new UnsentOffsetCommitRequest(
+                    offsets,
+                    groupState.groupId,
+                    groupState.groupInstanceId.orElse(null),
+                    groupState.generation);
+            unsentOffsetCommits.add(request);
+            return request.future();
+        }
+
+        /**
+         *  We want to avoid duplication when sending an {@link OffsetFetchRequest}. The following checks are done:
+         *  <li>1. dedupe against unsents: if a duplicated request was previously made, we chain the future</>
+         *  <li>2. dedupe against incompleted: If a duplicated request was sent but hasn't gotten a results, we chain
+         *  the future.</>
+         *
+         *  <p>If the request is new, we chain a call back to remove itself from the {@code inflightOffsetFetches}
+         *  upon completion.</>
+         */
+        private CompletableFuture<Map<TopicPartition, OffsetAndMetadata>> addOffsetFetchRequest(final UnsentOffsetFetchRequest request) {
+            Optional<UnsentOffsetFetchRequest> dupe =
+                    unsentOffsetFetches.stream().filter(r -> r.sameRequest(request)).findAny();
+            Optional<UnsentOffsetFetchRequest> inflight =
+                    inflightOffsetFetches.stream().filter(r -> r.sameRequest(request)).findAny();
+
+            if (dupe.isPresent() || inflight.isPresent()) {
+                log.info("Duplicated OffsetFetchRequest: " + request.requestedPartitions);
+                dupe.orElseGet(() -> inflight.get()).chainFuture(request.future);
+            } else {
+                // remove the request from the outbound buffer: inflightOffsetFetches
+                request.future.whenComplete((r, t) -> {
+                    if (!inflightOffsetFetches.remove(request)) {
```

Comment:
Just curious when do you think this could ever happen?

One case I was thinking is if the consumer is shutting down, but I also feel that when we shutdown a consumer, we should try to stop the background thread first, and then clear the buffer, so that the response would not be received while the buffer is cleared.

Reply:
I was actually thinking that if it is not expected for normal cases, maybe we should log it as WARN instead, or even throw a fatal error if we believe it should never happen.

For the case you described: in that case, the `drain` should not return any unsent request to the network client, and hence their requests would not be sent, and hence no responses should be received triggering this logic, right? So that seems still not normally happening?

Reply:
Agreed, let's log a warning for this case.